src/bls/bls_worker.cpp

```diff
@@ -369,19 +368,15 @@ struct VectorAggregator {
 
     void Start()
     {
-        std::vector<AggregatorType*> aggregators;
         for (size_t i = 0; i < vecSize; i++) {
             std::vector<const T*> tmp(count);
             for (size_t j = 0; j < count; j++) {
                 tmp[j] = &(*vecs[start + j])[i];
             }
 
-            auto aggregator = new AggregatorType(std::move(tmp), 0, count, parallel, workerPool, std::bind(&VectorAggregator::CheckDone, this, std::placeholders::_1, i));
-            // we can't directly start the aggregator here as it might be so fast that it deletes "this" while we are still in this loop
-            aggregators.emplace_back(aggregator);
-        }
-        for (auto agg : aggregators) {
-            agg->Start();
+            auto self(this->shared_from_this());
+            auto aggregator = std::make_shared<AggregatorType>(std::move(tmp), 0, count, parallel, workerPool, [self, i](const T& agg) {self->CheckDone(agg, i);});
+            aggregator->Start();
```

Comment:
Can you explain why you decided to start them in this loop as opposed to the way it was done before?

Reply:
Once an `Aggregator` is created, it can be started, there is no reason to delay it, but before there was a memory management issue that prevented starting the aggregator earlier: `this` (i.e. the `VectorAggregator`) was raw sent to the `Aggregator` together with a `doneCallback` which deletes `this`. Thus, the following scenario could happen: `we can't directly start the aggregator here as it might be so fast that it deletes "this" while we are still in this loop` (mentioned in a comment, line 380). Now that we are managing the memory of `this` using a shared_ptr (`VectorAggregator` is created using `make_shared`), `this` object will be alive as long as someone requires it.